python/ray/train/lightgbm/lightgbm_trainer.py

```diff
@@ -1,15 +1,223 @@
-from typing import Any, Dict, Union
+import logging
+from functools import partial
+from typing import Any, Dict, Optional, Union
 
 import lightgbm
 
+import ray
 from ray.train import Checkpoint
+from ray.train.constants import _DEPRECATED_VALUE, TRAIN_DATASET_KEY
 from ray.train.gbdt_trainer import GBDTTrainer
 from ray.train.lightgbm import RayTrainReportCallback
+from ray.train.lightgbm.v2 import LightGBMTrainer as SimpleLightGBMTrainer
+from ray.train.trainer import GenDataset
 from ray.util.annotations import PublicAPI
 
+logger = logging.getLogger(__name__)
 
+
+def _lightgbm_train_fn_per_worker(
+    config: dict,
+    label_column: str,
+    num_boost_round: int,
+    dataset_keys: set,
+    lightgbm_train_kwargs: dict,
+):
+    checkpoint = ray.train.get_checkpoint()
+    starting_model = None
+    remaining_iters = num_boost_round
+    if checkpoint:
+        starting_model = RayTrainReportCallback.get_model(checkpoint)
+        starting_iter = starting_model.current_iteration()
+        remaining_iters = num_boost_round - starting_iter
+        logger.info(
+            f"Model loaded from checkpoint will train for "
+            f"additional {remaining_iters} iterations (trees) in order "
+            "to achieve the target number of iterations "
+            f"({num_boost_round=})."
+        )
+
+    train_ds_iter = ray.train.get_dataset_shard(TRAIN_DATASET_KEY)
+    train_df = train_ds_iter.materialize().to_pandas()
+
+    eval_ds_iters = {
+        k: ray.train.get_dataset_shard(k)
+        for k in dataset_keys
+        if k != TRAIN_DATASET_KEY
+    }
+    eval_dfs = {k: d.materialize().to_pandas() for k, d in eval_ds_iters.items()}
+
+    # NOTE: Include the training dataset in the evaluation datasets.
+    # This allows `train-*` metrics to be calculated and reported.
+    eval_dfs[TRAIN_DATASET_KEY] = train_df
+
+    train_X, train_y = train_df.drop(label_column, axis=1), train_df[label_column]
+    train_set = lightgbm.Dataset(train_X, label=train_y)
+    valid_sets = []
+    for eval_df in eval_dfs.values():
+        eval_X, eval_y = eval_df.drop(label_column, axis=1), eval_df[label_column]
+        valid_sets.append(lightgbm.Dataset(eval_X, label=eval_y))
+    valid_names = list(eval_dfs.keys())
+
+    lightgbm.train(
+        params=config,
```

Comment:
Needs to include `get_network_params()`? 😄

Reply:
yes good catch!

Reply:
Make sure this actually runs distributed training 😏

Reply:
Yeah, I will need to add back a test suite like this one to make sure distributed training is actually happening properly.

https://github.com/ray-project/xgboost_ray/blob/5a840af05d487171883dadbfdd37b138b607bed8/xgboost_ray/tests/test_end_to_end.py#L56

For now I have confirmed it by running some examples.